### ISSUE TITLE
Record Yesterday's Activity (RYA) confusion: sync errors 

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica#8914]
_This is not exactly a bug but it's a behaviour that's confusing enough that it would be good to change it if we can. It's something to look at after the new website is released, if it's still happening then._

<strike>When you see the RYA modal first thing in the day, you can click Escape to dismiss it without going through the process of completing Dailies and starting your new day. You can keep doing this all day if you want to. When you finally do click the start button, cron runs but that can be hours after your CDS time, and can even be shortly before the CDS on the next day. This can cause confusion and can make players think cron is running at incorrect times or running twice a day. There's no actual bug because RYA and cron are performing as coded and the problem is that players are dismissing the modal instead of actioning it, but it's easy to do that by mistake first thing in the morning when you might have a few modals displaying at the same time and are in a hurry to get to your tasks.</strike> EDIT: That doesn't happen in the new site. 

----

**_EDIT 2018-07-24: The situation described below is still a problem. If the RYA modal opens in two tabs, after you've processed it in the first tab, if you later go to the second tab, you'll see the modal still open. If you don't understand what happens, you'll complete all your Dailies that you did_ yesterday _and then when you dismiss the modal, you'll find that the Dailies have been ticked for_ today. _That's correct behaviour from the API's point of view but confusing for users. I.e., when you start to use the RYA modal for the second time, there's no indication that you should not be using it now._**

If you have the website open in two tabs overnight, the modal can appear in both, leading you to be uncertain about whether you've already gone through cron, or fear that you might suffer a second cron that day (which Habitica's code does prevent, but players won't always know that). Below is a comment from Aerilita (3b724a41-48d3-4048-ae27-70b258cf5a72), describing it from a player's point of view:

"I'd like to report an issue related to the "start the new day" dialogue corresponding to the Habitica tab left open in the browser.So, yesterday was a good day and only two of my due dailies were left unchecked, and I didn't forget to use Stealth, so when I accessed Habitica today I was presented with the "start the new day" dialogue with those two dailies, left them unchecked, started my day. Got a finished quest popup, send an invitation to another one, checked a few dailies, then opened a new tab to do other stuff and forgot to close the Habitica tab, as I tend to do. It takes my brower maybe 30-40 minutes to "forget" the tab contents if the tab isn't accessed, and enought time has passed until I switched to that tab again. Then the browser kind of "reloaded" the page, but not actually (I don't know what it sends to Habitica, but not the full reload command). As the result, I was presented with the "start the new day" dialogue again, only now it had all of my dailies that were due today. (A similar thing happened a while before, and I think I actually clicked "start my new day" and sustained damage for undone dailies, but I was kind of hazy then, so I can't be sure I didn't just miss a day in casting Stealth.) As today I was absolutely sure I already performed the day-starting actions, I figured it's to do with the tab and not the site, so I just closed the tab and opened Habitica again, and this time it opened as it should.I remember someone reporting a similar issue with Bailey's news and other pop-up windows (everytime they return to the tab, they get those notifications) and a mod explained that this is due to the way browser loads the site for an already opened tab. But those were just confusing, while the "start the new day" is actually kind of scary when it shows up again, especially if one doesn't have a clear mental track of having used Habitica prior to that (and that's actually one of the things we need Habitica for, keeping track when we can't remember). Even scarier if pressing the "start my new day" button on the second time could trigger a premature cron (I don't know if it can, but I don't know if it can't). I don't know if this combination of browser behaviour and Habitica response could be fixed, but I hope someone can check conditions on that button.Just in case, first accessing Habitica today was around 10am, way past my custom dayroll time (which is 4am), so it shouldn't be related to that."

----

<strike>Another source of confusion related to RYA is that people will stop resting in the inn before they use the RYA screen, thinking that they have already avoided damage from cron. This is especially likely to happen for old users who were used to being able to check out of the inn first thing in the morning, because that action itself would trigger cron before the check out took effect, and they'd be safe.</strike> EDIT: That doesn't happen in the new site because you can't use the site after your CDS without RYA appearing. 

----

_**EDIT 2018-07-23: The situation described below still happens. If you click the "start my day" button before your clicks on the Dailies have been saved to the server, you take damage from Dailies that you had tried to mark as completed. It's a problem because users don't realise they need to slow down.**_

_**EDIT 2020-05-23: this part of this issue has been logged at https://github.com/HabitRPG/habitica/issues/11058**_

Sync errors are also more damaging in the RYA modal than normally because you're more likely to be clicking a few tasks quickly, and the effect of the sync error is immediate damage - no chance to notice that your actions haven't been saved before cron runs, if you go through the RYA modal quickly.

As an example of that (I think), here's a bug report that from r1243 (c2738c24-7889-4f66-857e-57e2b0f40f22) and my response:

"heya, just had a little glitch with the new 'tick-off-yesterday's-dailies-today' feature - opened my habitica tab just now, the screen popped up, ticked off my daily called duolingo and hit 'start next day', but I just took damage, my daily went redder and the daily's counter reset (so basically, it didn't count the click). I've been having similar issues with dailies not ticking off after a click/instantly turning back to unticked in general. running Chrome 59 on Windows 10 Professional, but it seems to have persisted across different computers. is this a known issue, and is there any way to avoid it happening?"

"It sounds like a [sync error](http://habitica.wikia.com/wiki/Sync_Errors). Sometimes when you take an action like marking a task as complete, the website (or mobile app) can't send your action to the server for permanent storage. The website/app does display the results of the action to you (e.g., it marks the task complete), but because the action wasn't saved to your account, it didn't really happen as far as Habitica is concerned. When the website/app later syncs with the server, it realises that the action wasn't saved and so it reverses the effects that it showed you (e.g., unticks the task). If this happens while you're using the Record Yesterday's Activity screen, then the affected Dailies won't be considered complete and you'll take damage from them.  Sync errors typically happen if your internet connection is weak or intermittent or if you take several actions in quick succession - the website/app can't send them all to the server fast enough to keep up with you.  Does it sound like that might be what's happening? When you're completing Dailies, are you often clicking several of them quickly? If so, can you try waiting a second or in between each click to see if that makes a difference?"

**Comment by librarianmage** [comment]
Should this be on hold?

**Comment by Alys** [comment]
I've removed the parts that are no longer relevant and marked this as suggestion-discussion instead of on hold. Thanks MathWhiz!

**Comment by TheHollidayInn** [comment]
Is this still being reported?

**Comment by Alys** [comment]
At least the final part of the top post is still happening - we had a report of that this morning. I've added a comment in bold italics to that part to indicate it's still relevant.

I've also added another comment to an earlier part to say I'll do some testing. Leave this as "needs reply" until I've written back here about that.

**Comment by Alys** [comment]
The earlier part is still happening too. I've adjusted my bold "edit" comment in the top post to indicate that.

**Comment by veeeeeee** [comment]
Having Habitica open in two tabs is an edge case that we aren't able to dedicate resources to right now.

However, some of the sync errors we're seeing result may be resolved by a change to sync RYA tasks only after clicking the Start Day button.

**Comment by paglias** [comment]
This is going to be much easier with https://github.com/HabitRPG/habitica/pull/11389 done, not sure if we're interested in finishing it

**Comment by Alys** [comment]
The latter part of this issue is logged at https://github.com/HabitRPG/habitica/issues/11058 and I've added a reference to paglias's comment above.

The former half of the issue still happens but we don't see any clear reports of it. I'd guess that if you use two tabs, you're more likely to be a power user who understands more about what's happening when it occurs. I'm closing this but we can reopen if we do find that people are still confused by RYA in two tabs at once.

**Comment by Alys** [comment]
We've had a new report of buggy RYA behaviour. I'm just noting this here so that we have it recorded, and can point people at it if anyone else reports the same problem. It doesn't require that this issue be reopened because we have https://github.com/HabitRPG/habitica/issues/12248 which will fix this problem.


@bifca (058cb4b9-268a-4072-9ee9-abaae8f24a2f) reported in the Report a Bug guild that sometimes when they complete Dailies in the RYA screen, the Dailies lose their streak as if they were NOT completed for the previous day but ARE completed for the current day.

I've confirmed that behaviour in one of the Dailies they reported as being affected (task ID f271150b-6753-4b70-aeed-ed60b9f822a8). These are entries from its history array:

```
"history": [
	...
	{
		"date": 1590233655147,	// 2020-05-23 11:34:15 UTC
		"value": 180.72902071829944
	},
	{
		"date": 1590344148306,	// 2020-05-24 18:15:48 UTC
		"value": 180.14921105630788
	},
	{
		"date": 1590344148859,	// 2020-05-24 18:15:48 UTC
		"value": 180.72902071829944
	}
],
```

The latter two entries are dated at the same time that their cron ran that day. As you can see, the first of those two entries decreased the value (i.e., the Daily was recorded by cron as being incomplete for the previous day), and then the second entry increased the value (the Daily was marked as complete, with the timing making it happen as cron ran so that the Daily remained complete after cron had finished).
